stack-graphs/src/serde.rs

```diff
@@ -0,0 +1,643 @@
+use crate::{arena::Handle, json::Filter};
+
+use serde::{Deserialize, Serialize};
+use thiserror::Error;
+
+#[derive(Serialize, Deserialize, PartialEq, Eq, Debug, Default)]
+pub struct StackGraph {
+    files: Files,
+    nodes: Nodes,
+    edges: Edges,
+}
+
+#[derive(Debug, Error, PartialEq, Eq)]
+pub enum Error {
+    #[error("failed to load file `{0}`")]
+    FileNotFound(String),
+
+    #[error("duplicate file `{0}`")]
+    FileAlreadyPresent(String),
+
+    #[error("failed to locate node `{0}` in graph")]
+    NodeNotFound(u32),
+
+    #[error("no file data for node `{0}`")]
+    NoFileData(u32),
+}
+
+impl StackGraph {
+    pub fn from_graph<'a>(graph: &crate::graph::StackGraph, filter: &'a dyn Filter) -> Self {
+        let files = graph.filter_files(filter);
+        let nodes = graph.filter_nodes(filter);
+        let edges = graph.filter_edges(filter);
+        Self {
+            files,
+            nodes,
+            edges,
+        }
+    }
+
+    pub fn load_into(&self, graph: &mut crate::graph::StackGraph) -> Result<(), Error> {
+        self.load_files(graph)?;
+        self.load_nodes(graph)?;
+        self.load_edges(graph)?;
+        Ok(())
+    }
+
+    fn load_files(&self, graph: &mut crate::graph::StackGraph) -> Result<(), Error> {
+        // we check if any of the files we are about to introduce, are
+        // already present in this graph, to avoid accidental merges
+        if let Some(f) = self
+            .files
+            .data
+            .iter()
+            .find(|f| graph.get_file(f.as_str()).is_some())
+        {
+            return Err(Error::FileAlreadyPresent(f.to_owned()));
+        }
+
+        for f in self.files.data.iter() {
+            graph.add_file(f.as_str()).unwrap();
+        }
+
+        Ok(())
+    }
+
+    fn load_nodes(&self, graph: &mut crate::graph::StackGraph) -> Result<(), Error> {
+        let not_found = |f: &str| Error::FileNotFound(f.to_owned());
+        let no_file_data = |id| Error::NoFileData(id);
+
+        for n in self.nodes.data.as_slice() {
+            match n {
+                Node::DropScopes { id, .. } => {
+                    let file = id.file().ok_or(no_file_data(id.local_id))?;
+                    let file_handle = graph.get_file(file).ok_or(not_found(file))?;
+                    let node_id = crate::graph::NodeID::new_in_file(file_handle, id.local_id);
```

Comment:
Node IDs should be constructed a little differently. Some node IDs without the file part are indeed valid. A function similar to this would do the trick:

```rust
impl NodeID {
  to_node_id(&self, graph: &crate::graph::StackGraph) -> Result<crate::graph::NodeID, Error> {
    let Some(file) = self.file {
      let file = graph.get_file(file).ok_or(not_found(file))?;
      Ok(NodeID::new_in_file(file, self.local_id))
    } else if self.local_id == ROOT_NODE_ID {
      Ok(NodeID::root())
    } else if self.local_id == JUMP_TO_NODE_ID {
      Ok(NodeID::jump_to())
    } else {
      Err(Error::InvalidNodeID(self))
    }
  }
}
```

Reply:
this reads a lot better, thanks! just to confirm, the only nodes that are permitted to not have a file (i.e., belong to all files) are the root and jump nodes, no?

Reply:
Correct.